aptos-move/aptos-gas-algebra/src/abstract_algebra.rs

```diff
@@ -0,0 +1,273 @@
+/// evaluated/materialized.
+pub trait GasExpression<E> {
+    type Unit;
+
+    /// Evaluates the expression within the given environment to a concrete number.
+    fn evaluate(&self, feature_version: u64, env: &E) -> GasQuantity<Self::Unit>;
+
+    /// Traverse the expression in post-order using the given visitor.
+    /// See [`GasExpressionVisitor`] for details.
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor);
+
+    /// Performs a division on the unit of the expression.
+    ///
+    /// This is sometimes required if you want to multiply an amount by a certain count.
+    fn per<U>(self) -> GasPerUnit<Self, U>
+    where
+        Self: Sized,
+    {
+        GasPerUnit {
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An interface for performing post-order traversal of the tree structure of a gas expression.
+///
+/// Alternatively, one could think that the callbacks are invoked following the Reverse Polish
+/// notation of the expression.
+///
+/// Here are a few examples:
+/// - `1 + 2`
+///   - `quantity(1)`
+///   - `quantity(2)`
+///   - `add()`
+/// - `A + B * 50`
+///   - `gas_param<A>()`
+///   - `gas_param<B>()`
+///   - `quantity(50)`
+///   - `mul()`
+///   - `add()`
+pub trait GasExpressionVisitor {
+    fn add(&mut self);
+
+    fn mul(&mut self);
+
+    fn gas_param<P>(&mut self);
+
+    fn quantity<U>(&mut self, quantity: GasQuantity<U>);
+
+    fn per<U>(&mut self);
+}
+
+/***************************************************************************************************
+ * Built-in Gas Expressions
+ *
+ **************************************************************************************************/
+/// Representing the addition of two gas expressions.
+#[derive(Debug, Clone)]
+pub struct GasAdd<L, R> {
+    pub left: L,
+    pub right: R,
+}
+
+/// Representing the multiplication of two gas expressions.
+#[derive(Debug, Clone)]
+pub struct GasMul<L, R> {
+    pub left: L,
+    pub right: R,
+}
+
+/// Representing a gas expression divided by a particular unit.
+/// This is sometimes required for further multiplications.
+#[derive(Debug, Clone)]
+pub struct GasPerUnit<T, U> {
+    pub inner: T,
+    phantom: PhantomData<U>,
+}
+
+/***************************************************************************************************
+ * Gas Expression Impl
+ *
+ **************************************************************************************************/
+// Notation:
+//   E | T: U means T is a valid gas expression with unit U under environment E.
+
+// E | T: U
+// ---------
+// E | &T: U
+impl<E, T> GasExpression<E> for &T
+where
+    T: GasExpression<E>,
+{
+    type Unit = T::Unit;
+
+    #[inline]
+    fn evaluate(&self, feature_version: u64, env: &E) -> GasQuantity<Self::Unit> {
+        (*self).evaluate(feature_version, env)
+    }
+
+    #[inline]
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor) {
+        (*self).visit(visitor)
+    }
+}
```

Comment:
I'm also planning to add a new type of node in the future, allowing tags to be attached to parts of an expression.

This has the potential to solve one problem I currently have with the gas profiler: the difficulty to break down the total storage fee into smaller items. I barely circumvented the problem by duplicating the logic, but there is huge risk that the gas meter & the profiler would go out of sync in the future, causing results to be inaccurate.